pkg/scheduler/queue/tenant_querier_assignment.go

```diff
@@ -358,3 +369,163 @@ func (tqa *tenantQuerierAssignments) shuffleTenantQueriers(tenantID TenantID, sc
+
+	// iterate through the tenant order until we find a tenant that is assigned to the current querier, or
+	// have checked the entire tenantIDOrder, whichever comes first
+	for iters := 0; iters < len(tqa.tenantIDOrder); iters++ {
+		if checkIndex >= len(tqa.tenantIDOrder) {
+			// do not use modulo to wrap this index; tenantOrderIndex is provided from an outer process
+			// which does not know if the tenant list has changed since the last dequeue
+			// wrapping with modulo after shrinking the list could cause us to skip tenants
+			checkIndex = 0
+		}
+		tenantID := tqa.tenantIDOrder[checkIndex]
+		tenantName := string(tenantID)
+
+		if _, ok := node.queueMap[tenantName]; !ok {
+			// tenant not in _this_ node's children, move on
+			checkIndex++
+			continue
+		}
+
+		checkedAllNodes = node.childrenChecked == len(node.queueMap)+1
+
+		// if the tenant-querier set is nil, any querier can serve this tenant
+		if tqa.tenantQuerierIDs[tenantID] == nil {
+			tqa.tenantOrderIndex = checkIndex
+			return node.queueMap[tenantName], checkedAllNodes
+		}
+		// otherwise, check if the querier is assigned to this tenant
+		if tenantQuerierSet, ok := tqa.tenantQuerierIDs[tenantID]; ok {
+			if _, ok := tenantQuerierSet[tqa.currentQuerier]; ok {
+				tqa.tenantOrderIndex = checkIndex
+				return node.queueMap[tenantName], checkedAllNodes
+			}
+		}
+		checkIndex++
+	}
+	return nil, checkedAllNodes
+}
+
+// dequeueUpdateState deletes the dequeued-from node from the following locations if it is empty:
+//   - parent's queueMap,
+//   - tenantNodes
+//   - tenantIDOrder iff there are no other nodes by the same name in tenantNodes. If the child is at the end of
+//     tenantIDOrder, it is removed outright; otherwise, it is replaced with an empty ("") element
+//
+// dequeueUpdateState would normally also handle incrementing the queue position after performing a dequeue, but
+// tenantQuerierAssignments currently expects the caller to handle this by having the querier set tenantOrderIndex.
+func (tqa *tenantQuerierAssignments) dequeueUpdateState(node *Node, dequeuedFrom *Node) {
+	// if dequeuedFrom is nil or is not empty, we don't need to do anything;
+	// position updates will be handled by the caller, and we don't need to remove any nodes.
+	if dequeuedFrom == nil || !dequeuedFrom.IsEmpty() {
+		return
+	}
+
+	// delete from the node's children
+	childName := dequeuedFrom.Name()
+	delete(node.queueMap, childName)
+
+	// delete from shared tenantNodes
+	for i, tenantNode := range tqa.tenantNodes[childName] {
+		if tenantNode == dequeuedFrom {
+			tqa.tenantNodes[childName] = append(tqa.tenantNodes[childName][:i], tqa.tenantNodes[childName][i+1:]...)
+		}
+	}
+
+	// check tenantNodes; we only remove from tenantIDOrder if all nodes with this name are gone.
+	// If the removed child is at the _end_ of tenantIDOrder, we remove it outright; otherwise,
+	// we replace it with an empty string so as not to reindex all slice elements
```

Comment:
> so as not to reindex all slice elements

I'd like to maintain some of the original comment's clarity on the fact that this is to not screw up how `lastTenantIndex` works. That way it's clearer that it can't be removed unless that approach is changed significantly - I think the "so as to not re-index" makes it sound more like an optimization.

original comment:
```
// Shrink tenant list if possible by removing empty tenant IDs.
// We remove only from the end; removing from the middle would re-index all tenant IDs
// and skip tenants when starting iteration from a querier-provided lastTenantIndex.
// Empty tenant IDs stuck in the middle of the slice are handled
// by replacing them when a new tenant ID arrives in the queue.
```